### ISSUE TITLE
Plugin registration silently fails when the Dart class name is already being used by another plugin

### ISSUE THREAD

**Issue by mateusfccp** [flutter/flutter#120391]
This issue raised from https://github.com/mateusfccp/update_available/issues/21. In my case, for some reason my plugin was simply not registering properly the Android implementation, so it fell back to `UnimplementedError`.

Renaming both iOS and Android implementations classes to be different fixed the issue.

## Steps to Reproduce

1. Create a federated plugin.
2. Create an implementation for Android and another for iOS.
3. In both implementation, use the same class name.
4. Setup the pubspecs properly to point to your implementations and everything.
5. Create an example that uses your plugin.
6. Run the example on Android.

**Expected results:** The plugin should work flawlessly.

**Actual results:**  When trying to access the supposedly registered instance, it actually fallsback to the unimplemented one, throwing an exception.

<details>
  <summary>Logs</summary>

```
pintome@BTGDEV548NMD6M ~/P/u/update_available (master)> flutter doctor -v
[✓] Flutter (Channel stable, 3.3.10, on macOS 13.1 22C65 darwin-x64, locale
    en-BR)
    • Flutter version 3.3.10 on channel stable at
      /Users/pintome/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 135454af32 (8 weeks ago), 2022-12-15 07:36:55 -0800
    • Engine revision 3316dd8728
    • Dart version 2.18.6
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/pintome/Library/Android
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/pintome/Library/Android/
    • Java binary at:
      /Library/Java/JavaVirtualMachines/temurin-11.jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment Temurin-11.0.14+9 (build
      11.0.14+9)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13F17a
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Users/pintome/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4345.14.2221.9189903
      /Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[!] Android Studio
    • Android Studio at /Users/pintome/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.4884.69.2231.9486165
      /Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[!] Android Studio
    • Android Studio at /Users/pintome/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9445173
      /Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.1 22C65 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 110.0.5481.77

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 3 categories.
```
</details>

**Comment by stuartmorgan** [comment]
> In my case, for some reason my plugin was simply not registering properly the Android implementation, so it fell back to `UnimplementedError`.
> 
> Renaming both iOS and Android implementations classes to be different fixed the issue.

The title here says registration "silently fails", but https://github.com/mateusfccp/update_available/pull/22#issuecomment-1424290773 says that it results in a compilation error, and suggests a much more likely explanation for lack of registration. Have you actually reproduced this?

> 2\. Create an implementation for Android and another for iOS.

In separate packages?

> When trying to access the supposedly registered instance, it actually fallsback to the unimplemented one, throwing an exception.

Please provide a test case, with the plugin correctly set up to do autoregistration, that demonstrates this.

**Comment by mateusfccp** [comment]
> The title here says registration "silently fails", but https://github.com/mateusfccp/update_available/pull/22#issuecomment-1424290773 says that it results in a compilation error, and suggests a much more likely explanation for lack of registration. Have you actually reproduced this?

Well, I can't answer for his case because I didn't get any error in compile-time. The behavior I had was as described in the this issue and the original bug report, an `UnimplementedError` in runtime that happened because the plugin was not properly registered (and thus, the static instance was not overriden).

> In separate packages?
Yes, the same structure that the federated plugin documentation suggests.

> Please provide a test case, with the plugin correctly set up to do autoregistration, that demonstrates this.
I am not sure how to make a test case for this, but I am going to try to make a minimum reproducible example in a separate repository.

**Comment by stuartmorgan** [comment]
> Well, I can't answer for his case because I didn't get any error in compile-time. The behavior I had was as described in the this issue and the original bug report, an `UnimplementedError` in runtime that happened because the plugin was not properly registered (and thus, the static instance was not overriden).

Which would be explained by leaving out `implements`, as the PR author said; as far as I can tell you didn't respond to that comment either in the PR or here. And based on that PR, you were in fact missing it.

I understand that you had a situation where autoregistration wasn't happening, my question is whether you actually reproduced a case where the plugin metadata **was correct**, and the cause of that situation was just having two classes with the same name (which is what this issue asserts) vs. a different cause as suggested by the discussion in that PR.

**Comment by mateusfccp** [comment]
> Which would be explained by leaving out `implements`, as the PR author said; as far as I can tell you didn't respond to that comment either in the PR or here. And based on that PR, you were in fact missing it.

I confess that I did not see that specific comment. However, long before the PR fixing the issue has been opened, I tried locally to fix it and was not successful. In my experiments, I made all the changes that the PR had except for the classes names, i.e I provided the missing `implements:` and ` dartPluginClass` in the `pubspec.yaml`. Even so, I did not get any error when building, only at runtime. If I did indeed get an error in compile time I would've fixed the issue by myself long before this PR has been opened.

I wonder if maybe my `.dart_tools` for some reason did not get rebuilt and it was the reason I was not getting any compile-time errors but runtime errors instead. It may be that although I provided the missing registration elements in the pubspec, the `.dart_tools` did not update. I can't remember if in my tests I use `flutter clean` to clean this kind of caches, so this is a possibility.

I am going to try to reproduce this issue again when I have some time. If I can't, I close this issue.
